docs(plugin-procfs): README modify (#245)
alumet-dev/alumet: plugins/procfs/README.md

# Procfs plugin

> Collects processes and system-related metrics by reading the [proc](<https://www.kernel.org/doc/html/latest/filesystems/proc.html>) virtual filesystem on Linux based operating systems.

Reviewer: ```suggestion
Collects processes and system-related metrics by reading the [proc](https://www.kernel.org/doc/html/latest/filesystems/proc.html) virtual filesystem on Linux based operating systems.
```

`<url>` is for bare urls. When you have `[label](url)`, don't put `<>`.

|`memory_usage`|Gauge|bytes|Memory usage|LocalMachine|Process|[kind](#kind)|

And in case we configure the plugin with the collect of detailed memory metrics :

|Name|Type|Unit|Description|Resource|ResourceConsumer|Attributes|
|----|----|----|-----------|--------|----------------|----------|
|`mem_total`|Gauge|bytes|Total amount of physical memory available in the system|LocalMachine|LocalMachine||
|`mem_free`|Gauge|bytes|Refers to truly free memory, which is not used by processes, caches or buffers|LocalMachine|LocalMachine||
|`mem_available`|Gauge|bytes|More accurate estimate of the actual available memory, taking into account free memory and also caches and buffers that can be freed|LocalMachine|LocalMachine||
|`cached`|Gauge|bytes|Cached is the memory used for ROM cache (page cache) in memory|LocalMachine|LocalMachine||
|`swap_cached`|Gauge|bytes|Memory that was placed in swap and brought back to RAM, but remains in swap file|LocalMachine|LocalMachine||
|`active`|Gauge|bytes|Memory that has been used recently|LocalMachine|LocalMachine||
|`inactive`|Gauge|bytes|Used memory that can be reclaimed, freed, or moved to swap.|LocalMachine|LocalMachine||

> |`mapped`|Gauge|bytes|Memory used for files associated with processes in memory, file and memory mapping.|LocalMachine|LocalMachine||

Reviewer: If this is about `plugins.procfs.memory.metrics`, this table lacks some values. But don't copy the whole doc, it's better to remove the table and explain that the metrics come from `/proc/meminfo`, with their names turned into snake_case (e.g. `MemTotal` becomes `mem_total`). Add a link to https://man7.org/linux/man-pages/man5/proc_meminfo.5.html.

### Attributes

#### Kind

The kind of the memory is the allocated memory space reserved by the system or the hardware (<https://man7.org/linux/man-pages/man5/proc_pid_status.5.html>) :

|Value|Description|
|-----|-----------|

> |`resident`|Resident set size, measured in pages (same as VmRSS in `/proc/<pid>/status`)|

Reviewer: The metric in Alumet is not in pages.

|`shared`|number of resident shared pages (i.e., backed by a file) (same as RssFile+RssShmem in `/proc/<pid>/status`)|
|`virtual`|Virtual memory size (same as VmSize in `/proc/<pid>/status`)|

The kind of the CPU time delta is the average CPU time spent by various tasks :

|Value|Description|
|-----|-----------|
|`user`|CPU time spent by the process in user mode|
|`system`|CPU time used by the system (Kernel)|
|`guest`|CPU time used by a guest virtual machine when a hypervisor virtualizes environments|

#### cpu_state

> The CPU states correspond the amount of time, measured in ticks, the CPU has been in specific states. These fields are measured in ticks because the underlying data from the kernel is measured in ticks (<https://docs.rs/procfs-core/latest/procfs_core/struct.CpuTime.html>) :

Reviewer: 1. The metric in Alumet is in `millisecond`: [see code](https://github.com/alumet-dev/alumet/blob/c3efb9d9f82dbe3c4c88900966e605076e442a5e/plugins/procfs/src/kernel.rs#L55-L56)
2. `cpu_state` is not an amount of time, it's an attribute that indicates the kind of cpu time that is measured

|Value|Description|
|-----|-----------|
|`user`|Ticks spent in user mode|
|`nice`|Ticks spent in user mode with low priority (nice)|
|`system`|Ticks spent in system mode|
|`idle`|Ticks spent in the idle state|
|`irq`|Ticks servicing interrupts|
|`softirq`|Ticks servicing soft interrupts|
|`steal`|Ticks of stolen time. Stolen time is the time spent in other operating systems when running in a virtualized environment.|
|`guest`|Ticks spent running a virtual CPU for guest operating systems under control of the linux kernel|
|`guest_nice`|Ticks spent running a niced guest|

## Configuration

> Here is a configuration example of the plugin. It's part of the ALUMET configuration file (eg: `alumet-config.toml`). Its configuration is composed of different sections and each of them can be enabled or disabled with the `enabled` parameter to be set to true or false.

Reviewer: ```suggestion
Here is a configuration example of the plugin. It is composed of different sections. Each section can be enabled or disabled with the `enabled` boolean parameter.
```

> To active the plugin to collect metrics relative to the kernel utilization :

Reviewer: Can you create one markdown subsection for each sub-config? `### Kernel metrics` for instance

poll_interval = "5s"
```

Moreover, you can collect more or less precise metrics on memory consumption, by setting the level of detail you want to extract from `/proc/meminfo` :

```toml
[plugins.procfs.memory]
# `true` to enable the monitoring of memory information.
enabled = true
# How frequently should the memory information be flushed to the rest of the pipeline.
poll_interval = "5s"
# The entry to parse from `/proc/meminfo`.
metrics = [
    "MemTotal",
    "MemFree",
    "MemAvailable",
    "Cached",
    "SwapCached",
    "Active",
    "Inactive",
    "Mapped",
]
```

To enable process monitoring, you need to set the metrics collect policy via a `strategy` :

- `watcher` : Default strategy of system watcher to collect new processes, whatever it may be.
- `event` : Set this parameter to collect the process that acts as an internal event of  ALUMET.

```toml
[plugins.procfs.processes]
# `true` to enable the monitoring of processes.
enabled = true
# Watcher refresh interval.
refresh_interval = "2s"
# `true` to watch for new processes, `false` to only react to ALUMET events.
strategy = "watcher"
```

> Also, you can monitor groups of processes, i.e. to the subdirectories in `/proc`, each subdirectory representing a process identified by its process ID, and collect metrics about them. In the group configuration, you can specify the PID or PPID to only monitor a process that has this PID or parent PID, and a character pattern to monitor the processes whose executable path matches this regex :

Reviewer: The user does not care how `/proc` is organized exactly IMO.

```toml
[[plugins.procfs.processes.groups]]

> # Only monitor the processes whose executable path matches this regex (No precision fixed in this case).

Reviewer: I think that `fix` is not the right verb to use here. What do you want to say?

exe_regex = ""
# How frequently should the processes information be refreshed.
poll_interval = "2s"
# How frequently should the processes information be flushed to the rest of the pipeline.
flush_interval = "4s"
```

## More information

### Procfs Access

> To grant the required access to retrieve all metrics properly by reading `/proc` filesystem, you need to configure the parameter [hidepid](<https://docs.kernel.org/filesystems/proc.html#mount-options>) by editing the configuration file `/proc/mounts`. This setting is a mount option for the `/proc` filesystem, that is used to control the visibility of processes to unprivileged users. In this way, it can define the access restriction to `/proc/<pid>/` directories, and therefore visibility of processes stats.

Reviewer: So, after investigating, is there a link to some capabilities or not?
